CLI: create output directory

Currently, if I do
```
aqua -i src/aqua -o src/generated
```
If `generated` doesn't exist, `aqua` will complain:
```
> aqua -i ./src -o ./src/generated

Aqua JS: node node_modules/@fluencelabs/aqua/aqua.js -m node_modules -i ./src -o ./src/generated
Aqua JS: 2021.09.09 12:27:28 [ERROR] There is no path './src/generated'
```
I think Aqua should create the output directory. I see several reasons for that:
1. All other compilers I know create output directory themselves (`tsc --outDir`, `scala -d`, `rustc --out-dir`)
2. It is rather tedious to work around this:
   1. Either I always keep empty `src/generated` directory in git
   2. Or I should prepend each call to aqua with `mkdir src/generated`

Also, here's an example of `tsc` creating the whole directory structure, `mkdir -p` style:
```
$ tsc --outDir 1/2/3/4/5/6/7/8/9/10
$ tree -d
.
└── 2
    └── 3
        └── 4
            └── 5
                └── 6
                    └── 7
                        └── 8
                            └── 9
                                └── 10
                                    └── generated
                                        └── aqua
```

--- a/aqua-src/import-empty.aqua
+++ b/aqua-src/import-empty.aqua
@@ -1,9 +1,9 @@
 -- import.aqua
 module Import.Test
-import foobar from "export.aqua"
+import foo from "export"
 
-use foo as f from "export.aqua" as Exp
+use foo as f from "export" as Exp
 
-use "export.aqua"
+use "export"
 
-export foobar as barfoo
+export foo as barfoo

--- a/aqua-src/import.aqua
+++ b/aqua-src/import.aqua
@@ -1,12 +1,9 @@
-service SomeService("erf"):
-  someCall()
+module Import
 
-func string_nil() -> *string:
-  value_nil: *string
-  SomeService.someCall()
-  <- value_nil
+use "export"
 
-func post() -> *string:
-  relay_nil: **string
-  relay_nil <- string_nil()
-  <- relay_nil
+export foo_wrapper
+
+func foo_wrapper() -> string:
+    z <- Export.foo()
+    <- z

--- a/build.sbt
+++ b/build.sbt
@@ -17,7 +17,7 @@ val declineV = "2.1.0"
 name := "aqua-hll"
 
 val commons = Seq(
-  baseAquaVersion := "0.3.0",
+  baseAquaVersion := "0.3.1",
   version         := baseAquaVersion.value + "-" + sys.env.getOrElse("BUILD_NUMBER", "SNAPSHOT"),
   scalaVersion    := dottyVersion,
   libraryDependencies ++= Seq(

--- a/build.sbt
+++ b/build.sbt
@@ -18,7 +18,7 @@ name := "aqua-hll"
 
 val commons = Seq(
   baseAquaVersion := "0.3.1",
-  version         := baseAquaVersion.value + "-" + sys.env.getOrElse("BUILD_NUMBER", "SNAPSHOT"),
+  version         := baseAquaVersion.value + "." + sys.env.getOrElse("BUILD_NUMBER", "SNAPSHOT"),
   scalaVersion    := dottyVersion,
   libraryDependencies ++= Seq(
     "com.outr"      %%% "scribe"    % "3.5.5",

--- a/cli/.jvm/src/main/scala/aqua/Test.scala
+++ b/cli/.jvm/src/main/scala/aqua/Test.scala
@@ -16,7 +16,7 @@ object Test extends IOApp.Simple {
       start <- IO(System.currentTimeMillis())
       _ <- AquaPathCompiler
         .compileFilesTo[IO](
-          Path("./aqua-src/import.aqua"),
+          Path("./aqua-src/"),
           List(Path("./aqua")),
           Path("./target"),
           TypeScriptBackend,

--- a/cli/.jvm/src/main/scala/aqua/Test.scala
+++ b/cli/.jvm/src/main/scala/aqua/Test.scala
@@ -20,7 +20,8 @@ object Test extends IOApp.Simple {
           List(Path("./aqua")),
           Path("./target"),
           TypeScriptBackend,
-          TransformConfig()
+          TransformConfig(),
+          false
         )
         .map {
           case Validated.Invalid(errs) =>

--- a/cli/.jvm/src/test/scala/WriteFileSpec.scala
+++ b/cli/.jvm/src/test/scala/WriteFileSpec.scala
@@ -20,7 +20,7 @@ class WriteFileSpec extends AnyFlatSpec with Matchers {
 
     val bc = TransformConfig()
     AquaPathCompiler
-      .compileFilesTo[IO](src, List.empty, targetTs, TypeScriptBackend, bc)
+      .compileFilesTo[IO](src, List.empty, targetTs, TypeScriptBackend, bc, false)
       .unsafeRunSync()
       .leftMap { err =>
         println(err)
@@ -32,7 +32,7 @@ class WriteFileSpec extends AnyFlatSpec with Matchers {
     Files[IO].deleteIfExists(targetTsFile).unsafeRunSync()
 
     AquaPathCompiler
-      .compileFilesTo[IO](src, List.empty, targetJs, JavaScriptBackend, bc)
+      .compileFilesTo[IO](src, List.empty, targetJs, JavaScriptBackend, bc, false)
       .unsafeRunSync()
       .leftMap { err =>
         println(err)
@@ -44,7 +44,7 @@ class WriteFileSpec extends AnyFlatSpec with Matchers {
     Files[IO].deleteIfExists(targetJsFile).unsafeRunSync()
 
     AquaPathCompiler
-      .compileFilesTo[IO](src, List.empty, targetAir, AirBackend, bc)
+      .compileFilesTo[IO](src, List.empty, targetAir, AirBackend, bc, false)
       .unsafeRunSync()
       .leftMap { err =>
         println(err)

--- a/cli/src/main/scala/aqua/AppOps.scala
+++ b/cli/src/main/scala/aqua/AppOps.scala
@@ -93,7 +93,8 @@ object AppOps {
 
   def importOpts[F[_]: Monad: Files]: Opts[F[ValidatedNec[String, List[Path]]]] =
     Opts
-      .options[String]("import", "Path to the directory to import from", "m").orEmpty
+      .options[String]("import", "Path to the directory to import from", "m")
+      .orEmpty
       .map { ps =>
         val checked: List[F[ValidatedNec[String, Path]]] = ps.toList.map { pStr =>
           val p = Path(pStr)
@@ -114,7 +115,11 @@ object AppOps {
 
   def constantOpts[F[_]: LiftParser: Comonad]: Opts[List[TransformConfig.Const]] =
     Opts
-      .options[String]("const", "Constant that will be used in an aqua code. Constant name must be upper cased.", "c")
+      .options[String](
+        "const",
+        "Constant that will be used in an aqua code. Constant name must be upper cased.",
+        "c"
+      )
       .mapValidated { strs =>
         val parsed = strs.map(s => ConstantExpr.onlyLiteral.parseAll(s))
 
@@ -130,7 +135,7 @@ object AppOps {
                 TransformConfig.Const(v._1.value, LiteralModel(v._2.value, v._2.ts))
             })
           ) { errors =>
-            val errorMsgs = errors.map (str => s"Invalid constant definition '$str'.")
+            val errorMsgs = errors.map(str => s"Invalid constant definition '$str'.")
             Validated.invalid(errorMsgs)
           }
       }
@@ -157,6 +162,12 @@ object AppOps {
   val noXorWrapper: Opts[Boolean] =
     Opts
       .flag("no-xor", "Do not generate a wrapper that catches and displays errors")
+      .map(_ => true)
+      .withDefault(false)
+
+  val dryOpt: Opts[Boolean] =
+    Opts
+      .flag("dry", "Generate .air file instead of typescript")
       .map(_ => true)
       .withDefault(false)
 

--- a/cli/src/main/scala/aqua/AppOps.scala
+++ b/cli/src/main/scala/aqua/AppOps.scala
@@ -44,6 +44,23 @@ object AppOps {
       )
   }
 
+  def checkOutput[F[_]: Monad: Files](pathStr: String): F[ValidatedNec[String, Path]] = {
+    val p = Path(pathStr)
+    Files[F]
+      .exists(p)
+      .flatMap { exists =>
+        if (exists)
+          Files[F].isRegularFile(p).map { isFile =>
+            if (isFile) {
+              Validated.invalidNec(s"Output path should be a directory. Current: '$p'")
+            } else
+              Validated.validNec(p)
+          }
+        else
+          Files[F].createDirectories(p).map(_ => Validated.validNec(p))
+      }
+  }
+
   def checkPath[F[_]: Monad: Files](pathStr: String): F[ValidatedNec[String, Path]] = {
     val p = Path(pathStr)
     Files[F]
@@ -72,7 +89,7 @@ object AppOps {
       .map(s => checkPath[F](s))
 
   def outputOpts[F[_]: Monad: Files]: Opts[F[ValidatedNec[String, Path]]] =
-    Opts.option[String]("output", "Path to the output directory", "o").map(s => checkPath[F](s))
+    Opts.option[String]("output", "Path to the output directory", "o").map(s => checkOutput[F](s))
 
   def importOpts[F[_]: Monad: Files]: Opts[F[ValidatedNec[String, List[Path]]]] =
     Opts

--- a/cli/src/main/scala/aqua/AppOps.scala
+++ b/cli/src/main/scala/aqua/AppOps.scala
@@ -89,7 +89,7 @@ object AppOps {
       .map(s => checkPath[F](s))
 
   def outputOpts[F[_]: Monad: Files]: Opts[F[ValidatedNec[String, Path]]] =
-    Opts.option[String]("output", "Path to the output directory", "o").map(s => checkOutput[F](s))
+    Opts.option[String]("output", "Path to the output directory. Will be created if not exists", "o").map(s => checkOutput[F](s))
 
   def importOpts[F[_]: Monad: Files]: Opts[F[ValidatedNec[String, List[Path]]]] =
     Opts

--- a/cli/src/main/scala/aqua/AquaCli.scala
+++ b/cli/src/main/scala/aqua/AquaCli.scala
@@ -57,9 +57,10 @@ object AquaCli extends IOApp with Logging {
       wrapWithOption(helpOpt),
       wrapWithOption(versionOpt),
       logLevelOpt,
-      constantOpts[Id]
+      constantOpts[Id],
+      dryOpt
     ).mapN {
-      case (inputF, importsF, outputF, toAir, toJs, noRelay, noXor, h, v, logLevel, constants) =>
+      case (inputF, importsF, outputF, toAir, toJs, noRelay, noXor, h, v, logLevel, constants, isDryRun) =>
         scribe.Logger.root
           .clearHandlers()
           .clearModifiers()
@@ -91,7 +92,8 @@ object AquaCli extends IOApp with Logging {
                       imports,
                       output,
                       targetToBackend(target),
-                      bc
+                      bc,
+                      isDryRun
                     )
                   }
                 }

--- a/cli/src/main/scala/aqua/AquaPathCompiler.scala
+++ b/cli/src/main/scala/aqua/AquaPathCompiler.scala
@@ -1,7 +1,7 @@
 package aqua
 
 import aqua.backend.Backend
-import aqua.compiler.{AquaCompiler, AquaError}
+import aqua.compiler.{AquaCompiled, AquaCompiler, AquaError}
 import aqua.files.{AquaFileSources, FileModuleId}
 import aqua.io.*
 import aqua.model.transform.TransformConfig
@@ -9,8 +9,9 @@ import aqua.parser.{Ast, LexerError}
 import aqua.parser.lift.FileSpan
 import cats.data.*
 import cats.syntax.functor.*
+import cats.syntax.applicative.*
 import cats.syntax.show.*
-import cats.{~>, Eval, Monad, Show}
+import cats.{Eval, Monad, Show, ~>, Applicative}
 import scribe.Logging
 import fs2.io.file.{Files, Path}
 import aqua.parser.lift.{LiftParser, Span}
@@ -27,7 +28,8 @@ object AquaPathCompiler extends Logging {
     imports: List[Path],
     targetPath: Path,
     backend: Backend,
-    bodyConfig: TransformConfig
+    bodyConfig: TransformConfig,
+    isDryRun: Boolean
   ): F[ValidatedNec[String, Chain[String]]] = {
     import ErrorRendering.showError
     val sources = new AquaFileSources[F](srcPath, imports)
@@ -50,9 +52,16 @@ object AquaPathCompiler extends Logging {
         },
         backend,
         bodyConfig,
-        sources.write(targetPath)
+        {if (isDryRun) dry[F] else sources.write(targetPath)}
       )
       .map(_.leftMap(_.map(_.show)))
   }
+
+  def dry[F[_]: Applicative](ac: AquaCompiled[FileModuleId]): F[Seq[Validated[AquaFileError, String]]] =
+    Seq(
+      Validated.valid[AquaFileError, String](
+        s"Source ${ac.sourceId.file}: compilation OK"
+      )
+    ).pure[F]
 
 }

--- a/cli/src/main/scala/aqua/files/AquaFileSources.scala
+++ b/cli/src/main/scala/aqua/files/AquaFileSources.scala
@@ -56,8 +56,12 @@ class AquaFileSources[F[_]: AquaIO: Monad: Files: Functor](
     val validatedPath = Validated.fromEither(Try(Path(imp)).toEither.leftMap(FileSystemError.apply))
     validatedPath match {
       case Validated.Valid(importP) =>
+        // if there is no `.aqua` extension, than add it
+        val renamedImport =
+          if (importP.extName != ".aqua") importP.resolveSibling(importP.fileName.toString + ".aqua")
+          else importP
         filesIO
-          .resolve(importP, importFrom.prependedAll(from.file.parent))
+          .resolve(renamedImport, importFrom.prependedAll(from.file.parent))
           .bimap(NonEmptyChain.one, FileModuleId(_))
           .value
           .map(Validated.fromEither)

